src/modalities/nn/attention.py

```diff
@@ -0,0 +1,98 @@
+import math
+from enum import Enum
+from typing import Optional, Tuple
+
+import torch
+import torch.nn.functional as F
+from pydantic import BaseModel
+from torch import Tensor, nn
+
+
+class AttentionEngineType(str, Enum):
+    DEFAULT_ATTENTION = "default_attention"
+    PYTORCH_FLASH_ATTENTION = "pytorch_flash_attention"
+
+
+class AttentionType(str, Enum):
+    CAUSAL_ATTENTION = "causal_attention"
+    NON_CAUSAL_ATTENTION = "non_causal_attention"

```

Comment:
Maybe rename `causal_attention` to `causal_self_attention` and `non_causal_attention` to `non_causal_self_attention` in order to emphasize the distinction between self-attention and `cross_attention`?